crates/vm/levm/src/hooks/l2_hook.rs

```diff
@@ -1,31 +1,49 @@
 use crate::{
-    errors::{ContextResult, InternalError, TxValidationError, VMError},
+    errors::{ContextResult, InternalError, TxValidationError},
     hooks::{default_hook, hook::Hook},
+    opcodes::Opcode,
     vm::VM,
 };
 
-use ethrex_common::{Address, U256, types::Fork};
+use ethrex_common::{Address, H160, U256, types::Fork};
 
-pub struct L2Hook {
-    pub recipient: Option<Address>,
-}
+pub const COMMON_BRIDGE_L2_ADDRESS: Address = H160([
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0xff, 0xff,
+]);
+
+pub struct L2Hook {}
 
 impl Hook for L2Hook {
     fn prepare_execution(&mut self, vm: &mut VM<'_>) -> Result<(), crate::errors::VMError> {
-        if vm.env.is_privileged {
-            let Some(recipient) = self.recipient else {
-                return Err(InternalError::RecipientNotFoundForPrivilegedTransaction.into());
-            };
-            vm.increase_account_balance(recipient, vm.current_call_frame()?.msg_value)?;
-            vm.current_call_frame_mut()?.msg_value = U256::from(0);
-        }
-
         let sender_address = vm.env.origin;
         let (sender_balance, sender_nonce) = {
             let sender_account = vm.db.get_account(sender_address)?;
             (sender_account.info.balance, sender_account.info.nonce)
         };
 
+        // The bridge is allowed to mint ETH.
+        // This is done by not decreasing it's balance when it's the source of a transfer.
+        // For other privileged transactions, insufficient balance can't cause an error
+        // since they must always be accepted, and an error would mark them as invalid
+        // Instead, we make them revert by inserting a revert2
+        if vm.env.is_privileged && sender_address != COMMON_BRIDGE_L2_ADDRESS {
+            let value = vm.current_call_frame()?.msg_value;
+            if value > sender_balance {
+                vm.current_call_frame_mut()?.msg_value = U256::zero();
+                vm.current_call_frame_mut()?
+                    .set_code(vec![Opcode::INVALID.into()].into())?;
+                return Ok(());
```

Comment:
This early return might break the gas calculation and make it go over the block gas limit.